Lib/Inc/CPIDLoop.h

```diff
@@ -0,0 +1,97 @@
+/**
+ * @file CPIDLoop.h
+ */
+
+/*
+ * CPIDLoop.h
+ *
+ *  Created on: 19 Feb 2022
+ *      Author: salavat.magazov
+ */
+
+#ifndef CPIDLOOP_H_
+#define CPIDLOOP_H_
+
+/* Default PID parameters.
+ * These are setup to make PID loop transparent. Output will be equal to error.
+ * Proportional is percentage of power per unit of error.
+ * Integral is units of output per unit of error per unit of time.
+ * Differential is unit of output per degree error change per unit of time.
+ * Normally for slow moving processes differential is zero.
+ */
+#define DEFAULT_P_COEFF 1
+#define DEFAULT_I_COEFF 0
+#define DEFAULT_D_COEFF 0
+
+class CPIDLoop
+{
+public:
+    struct CONTROL_STRUCT_T
+    {
+        CONTROL_STRUCT_T(float k_p, float k_i, float k_d);
+
+        float p_coeff;
+        float i_coeff;
+        float d_coeff;
+
+        float min_out;
+        float max_out;
+
+        float min_i;
+        float max_i;
+
+        float min_d;
+        float max_d;
+    };
+
+    CPIDLoop(float k_p = DEFAULT_P_COEFF,
+             float k_i = DEFAULT_I_COEFF,
+             float k_d = DEFAULT_D_COEFF);
+
+    float run(float target, float actual);
+    float run(float target, float actual, float deltaError);
+    void reset();
+
+    /* PID runtime terms access methods. */
+    float getP() const
+    {
+        return m_p;
+    }
+
+    float getI() const
+    {
+        return m_i;
+    }
+    float getD() const
+    {
+        return m_d;
+    }
+    void setI(float i)
+    {
+        m_i = i;
+    }
+
+    void setupLoop(const CONTROL_STRUCT_T &control_struct)
```

Comment:
Is this the only way to change the coeff. limits? If I create a CPIDLoop instance with some kp, ki and kd values. Do I need to repeat them in the struct to change the default limits? 
```
CPIDLoop control_1(1,0,0); 
control_1.setupLoop(control_struct);  -> This control_struct needs to repeat 1,0,0?
```
Wouldn't it be easier to pass the reference to the control_struct as parameter in the CPIDLoop constructor?

Reply:
You do not need to create control_struct_t instance yourself. CPIDLoop will do that during instantiation.